### ISSUE TITLE
[ORACLE] Can't import vector layer to oracle database

### ISSUE THREAD

**Issue by ItaloCunhaBueno** [qgis/QGIS#39855]
Using DB Manager to import a vector layer to a oracle database gives the following error:

```
Error 2
Creation of data source "TESTE22S"."LOTE" failed: 
Oracle error: Could not insert metadata.
SQL: ORA-00932: inconsistent datatypes: expected MDSYS.SDO_DIM_ARRAY got CHAR
Unable to execute statement
Error: INSERT INTO mdsys.user_sdo_geom_metadata(table_name,column_name,srid,diminfo) VALUES (:a,:bb,:bc,:bd)
```

My config:

```
QGIS version
3.10.11-A Coruña
QGIS code revision
d2171173e4
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.1.4
Running against GDAL/OGR
3.1.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 10 (10.0)
```

Any help are welcome. Thanks.

**Comment by gioman** [comment]
> SQL: ORA-00932: inconsistent datatypes: expected MDSYS.SDO_DIM_ARRAY got CHAR
> Unable to execute statement

@ItaloCunhaBueno you sure is not an issue with your data/tables?

**Comment by ItaloCunhaBueno** [comment]
That's a system table, created automatically by oracle, if it's expecting SDO_DIM_ARRAY and QGIS try to passCHAR, I think it's a QGIS problem.

**Comment by gioman** [comment]
> 3.10.11-A Coruña

@ItaloCunhaBueno can ypu check also on QGIS 3.16 and master?

**Comment by ItaloCunhaBueno** [comment]
Tested on 3.16, same error. Can't test on master as is too much trouble to build from source.

Aditional information:
The table that QGIS is having trouble inserting is described in this page:

https://docs.oracle.com/database/121/SPATL/geometry-metadata-views.htm#SPATL545

the particularly value that QGIS is doing wrong is described in section "2.8.3 DIMINFO".

**Comment by ItaloCunhaBueno** [comment]
Here's an example of a valid row in this table:

```
"TABLE_NAME"	"COLUMN_NAME"	"DIMINFO"	"SRID"
--------------------------------------------------------------------------------------------------------------------
"CONFRONTANTES"	"GEOMETRY"	MDSYS.SDO_DIM_ARRAY(MDSYS.SDO_DIM_ELEMENT('X', -2147483648, 2147483647, 0.00005), MDSYS.SDO_DIM_ELEMENT('Y', -2147483648, 2147483647, 0.00005), MDSYS.SDO_DIM_ELEMENT('Z', -2147483648, 2147483647, 0.00005))	(null)
```

**Comment by gioman** [comment]
> Can't test on master as is too much trouble to build from source.

@ItaloCunhaBueno you find qgis master in the osgeo4w installer.

**Comment by rduivenvoorde** [comment]
I know what the problem is, 

This lines https://github.com/qgis/QGIS/blob/master/src/providers/oracle/qgsoracleprovider.cpp#L3096-L3097 fail because the mdsys.sdo_dim_array is inserted as a string instead of an object because of the use of the bind variables...

will do PR soon

**Comment by ItaloCunhaBueno** [comment]
> I know what the problem is,
> 
> This lines https://github.com/qgis/QGIS/blob/master/src/providers/oracle/qgsoracleprovider.cpp#L3096-L3097 fail because the mdsys.sdo_dim_array is inserted as a string instead of an object because of the use of the bind variables...
> 
> will do PR soon

Thank you very much!

And thank you @gioman for your time!

**Comment by ItaloCunhaBueno** [comment]
> I know what the problem is,
> 
> This lines https://github.com/qgis/QGIS/blob/master/src/providers/oracle/qgsoracleprovider.cpp#L3096-L3097 fail because the mdsys.sdo_dim_array is inserted as a string instead of an object because of the use of the bind variables...
> 
> will do PR soon

Just to be sure, in wich version should I expect this fix to be included?

**Comment by rduivenvoorde** [comment]
@ItaloCunhaBueno I'm busy in this Pull Request: https://github.com/qgis/QGIS/pull/39846
It turned out there were more things not happy in the Oracle Provider.
And even after fixing the error which generated you messages, I'm still not able to actually import data.
After the fixes in the pull, I can at least use drag/drop in the browser.
This will probably land in 3.18 and before that it will be available in the so called nightly builds (AFTER somebody accepted/tested my Pull Request).
See also https://qgis.nl/2020/10/03/institutional-centralized-qgis-installation-configuration-qgis-de/?lang=en the item about Testing "QGIS fixes on Windows" you can actually download the so called 'artifact' (==QGIS) from the Windows test build. Although it has some limitations (one of them I think is that it cannot run Python... which you need I think to run the db manager. But you can maybe try?

**Comment by ItaloCunhaBueno** [comment]
> See also https://qgis.nl/2020/10/03/institutional-centralized-qgis-installation-configuration-qgis-de/?lang=en the item about Testing "QGIS fixes on Windows" you can actually download the so called 'artifact' (==QGIS) from the Windows test build. Although it has some limitations (one of them I think is that it cannot run Python... which you need I think to run the db manager. But you can maybe try?

That's really awesome that one can do that. Thanks for all your help.

**Comment by ItaloCunhaBueno** [comment]
Saddly, this artifact don't come with oracle provider, so, I'm not even able to connect from de browser.

**Comment by rduivenvoorde** [comment]
If your really have to load some data, you can try to do it via ogr2ogr, which is in the osgeo4w (and I think also in the QGIS) distribution. See https://gdal.org/drivers/vector/oci.html For what I know it has Oracle compiled into gdal.

**Comment by alexbruy** [comment]
Is this still an issue in recent QGIS versions? Seems should be fixed by #40107.

**Comment by ItaloCunhaBueno** [comment]
The company I work don't use Oracle databases anymore, so I can't say for sure. I think it's safe to be closed.
